### PR TITLE
Use BoundInference in nonlinear extension

### PR DESCRIPTION
**Pull request by nafur**
Currently the NonlinearExtensions uses a custom logic to eliminate redundant bounds and perform tightening on bound integer terms. As these replacements are not recorded, incorrect conflicts are being sent to the InferenceManager.
This PR replaces this logic by the BoundInference class and fixes the issues with conflicts by
- allowing BoundInference to collect bounds on arbitrary left hand sides (instead of only variables),
- improving origin tracking in BoundInference by explicitly constructing the new bound constraints,
- adding tightening of integer bounds,
- emitting lemmas instead of conflicts, and finally
- replacing the current logic by using the BoundInference class.

### CODE REVIEW

**Comment by nafur** [comment]
Benchmarks across QF logics that involve nonlinear reasoning mostly improve:
```
$ ./analyze.sh -G slurmwd/10280045-all-master/ slurmwd/10280244-all-nlbi/
                              status  result  solved   sat  unsat  best  timeout  memout  error  uniq  disagr  time_cpu    memory
benchmark config                                                                                                                 
QF_ANIA   10280045-all-master     ok      95      68    66      2    65       27       0      0     0       0     872.8    3297.7
          10280244-all-nlbi       ok      95      68    66      2     3       27       0      0     0       0     872.3    3304.2
QF_AUFNIA 10280045-all-master     ok      17       7     3      4     2       10       0      0     0       0     336.0     995.3
          10280244-all-nlbi       ok      17       7     3      4     5       10       0      0     0       0     334.4     992.0
QF_NIA    10280045-all-master     ok   23876    7165  4891   2274  5708    10833       0      0   141       0  395410.6  891799.3
          10280244-all-nlbi       ok   23876    7217  4942   2275  1650    10746       0      0   193       0  394249.3  886726.3
QF_NIRA   10280045-all-master     ok       3       1     0      1     1        2       0      0     0       0      60.4     237.6
          10280244-all-nlbi       ok       3       1     0      1     0        2       0      0     0       0      60.4     267.2
QF_NRA    10280045-all-master     ok   11489   10052  4851   5201  9414     1438       0      0    40       0   49781.1  191250.5
          10280244-all-nlbi       ok   11489   10058  4858   5200   684     1432       0      0    46       0   49337.8  193208.8
QF_UFNIA  10280045-all-master     ok     478     436   266    170   416       12       0      0     9       0     521.7    2325.2
          10280244-all-nlbi       ok     478     430   263    167    23       14       0      0     3       0     613.8    2385.2
QF_UFNRA  10280045-all-master     ok      37      13     2     11    13       24       0      0     0       0     725.9    1051.3
          10280244-all-nlbi       ok      37      13     2     11     0       24       0      0     0       0     726.0    1059.8

$ ./analyze.sh --show-diff -G slurmwd/10280045-all-master/ slurmwd/10280244-all-nlbi/
                              status  result  solved  sat  unsat  best  timeout  memout  error  uniq  disagr  time_cpu   memory
benchmark config                                                                                                               
QF_NIA    10280045-all-master     ok     334     141  119     22   141      107       0      0   141       0    6110.9  10884.6
          10280244-all-nlbi       ok     334     193  170     23   193       82       0      0   193       0    5808.4  10623.1
QF_NRA    10280045-all-master     ok      86      40   28     12    40       46       0      0    40       0    1855.0   2493.9
          10280244-all-nlbi       ok      86      46   35     11    46       41       0      0    46       0    1651.5   2818.5
QF_UFNIA  10280045-all-master     ok      12       9    6      3     9        2       0      0     9       0      84.4    180.7
          10280244-all-nlbi       ok      12       3    3      0     3        3       0      0     3       0     129.9    183.3
```
Note the slight regression on `QF_UFNIA` which I still investigate. I observed that the new solution sometimes removes more redundancies than the old one. Instances that are lost seem all to be due to the CAD variable ordering changing...
Benchmarks on quantified logics are underway...

File: src/theory/arith/bound_inference.h [hunk 1]
```diff
@@ -27,65 +27,73 @@ namespace arith {
...
   /**
-   * Add a new theory atom. Return true if the theory atom induces a new
-   * variable bound.
+   * A utility class that extracts direct bounds on arithmetic terms from theory
+   * atoms.
    */
-  bool add(const Node& n);
+  class BoundInference
+  {
+    /** The currently strictest bounds for every lhs. */
```
Comment (ajreynol) [inline src/theory/arith/bound_inference.h:57]: I missed this previously, but these 3 things should go in a private block at the bottom of the class.

File: src/theory/arith/nl/nonlinear_extension.h [hunk 2]
```diff
@@ -250,6 +250,9 @@ class NonlinearExtension
    * and for establishing when we are able to answer "SAT".
    */
   NlModel d_model;
+
+  BoundInference d_boundInference;
```
Comment (ajreynol) [inline src/theory/arith/nl/nonlinear_extension.h:254]: Document, in particular when it is valid and what bounds it stores (for what?).

File: src/theory/arith/inference_manager.h [hunk 3]
```diff
@@ -84,7 +86,10 @@ class InferenceManager : public InferenceManagerBuffered
   void clearWaitingLemmas();
 
   /** Add a conflict to the this inference manager. */
-  void addConflict(const Node& conf, InferenceId inftype);
+  void addConflict(std::vector<Node>& conf, InferenceId inftype);
+
+  /** Add a conflict processor. */
```
Comment (ajreynol) [inline src/theory/arith/inference_manager.h:91]: This interface is not clear to me.  Who should call this? Does the conflict processor apply to all conflicts?  

Do we need this interface to be this general?

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/theory/arith/bound_inference.h [hunk 4]
```diff
@@ -27,66 +27,75 @@ namespace arith {
...
+
+    /**
+     * Add a new theory atom. Return true if the theory atom induces a new
+     * term bound.
+     * If onlyVariables is true, the left hand side needs to be a single
+     * variable to induce a bound.
+     */
+    bool add(const Node& n, bool onlyVariables = true);
+
+    void replaceByOrigins(std::vector<Node>& nodes) const;
```
Comment (ajreynol) [inline src/theory/arith/bound_inference.h:82]: Document this.

**Review by ajreynol — CHANGES_REQUESTED** [review]
One last documentation request.

Also, I believe the numbers are updated after the conflict -> lemma change.  Post these?

**Comment by nafur** [comment]
Current results:
```
$ ./analyze.sh -G slurmwd/10290801-all*
                                 status  result  solved   sat  unsat  best  timeout  memout  error  uniq  disagr  time_cpu    memory
benchmark   config                                                                                                                  
AUFBVDTNIA  10290801-allq-master     ok       1       0     0      0     0        0       0      0     0       0       0.0       2.1
            10290801-allq-nlbi       ok       1       0     0      0     0        0       0      0     0       0       0.0       2.1
AUFDTNIRA   10290801-allq-master     ok     471     327     0    327   280       57       0      0     0       0    1803.2    7191.9
            10290801-allq-nlbi       ok     471     327     0    327    47       57       0      0     0       0    1800.3    7229.1
AUFFPDTNIRA 10290801-allq-master     ee      13       0     0      0     0        0       0     13     0       0       0.0      22.7
            10290801-allq-nlbi       ee      13       0     0      0     0        0       0     13     0       0       0.0      22.5
AUFNIA      10290801-allq-master     ok       3       0     0      0     0        0       0      0     0       0       0.0       8.1
            10290801-allq-nlbi       ok       3       0     0      0     0        0       0      0     0       0       0.0       8.6
AUFNIRA     10290801-allq-master     ok    1480    1028     0   1028  1004      406       0      0     0       0   12507.8   54962.7
            10290801-allq-nlbi       ok    1480    1028     0   1028    24      405       0      0     0       0   12485.9   55283.7
NIA         10290801-allq-master     ok      20      18    11      7    17        0       0      0     0       0      12.7      89.3
            10290801-allq-nlbi       ok      20      18    11      7     1        0       0      0     0       0      12.8      88.0
NRA         10290801-allq-master     ok    3813    3782     0   3782  3780       26       0      0     0       0     817.8   10772.1
            10290801-allq-nlbi       ok    3813    3782     0   3782     2       26       0      0     0       0     819.4   10756.7
QF_ANIA     10290801-all-master      ok      95      68    66      2    66       27       0      0     0       0     868.4    2981.4
            10290801-all-nlbi        ok      95      68    66      2     2       27       0      0     0       0     876.2    3836.1
QF_AUFNIA   10290801-all-master      ok      17       7     3      4     6       10       0      0     0       0     334.3     997.2
            10290801-all-nlbi        ok      17       7     3      4     1       10       0      0     0       0     335.3     995.2
QF_NIA      10290801-all-master      ok   23876    7177  4904   2273  5859    10819       0      0   136       0  394645.4  893365.5
            10290801-all-nlbi        ok   23876    7228  4948   2280  1505    10734       0      0   187       0  393865.8  882935.3
QF_NIRA     10290801-all-master      ok       3       1     0      1     1        2       0      0     0       0      60.3     236.2
            10290801-all-nlbi        ok       3       1     0      1     0        2       0      0     0       0      60.4     233.5
QF_NRA      10290801-all-master      ok   11489   10057  4854   5203  9388     1432       0      0    29       0   49714.8  191343.0
            10290801-all-nlbi        ok   11489   10080  4866   5214   721     1409       0      0    52       0   48920.6  192376.9
QF_UFNIA    10290801-all-master      ok     478     436   266    170   413       12       0      0     9       0     520.0    2318.7
            10290801-all-nlbi        ok     478     430   263    167    26       14       0      0     3       0     609.1    2383.9
QF_UFNRA    10290801-all-master      ok      37      13     2     11    13       24       0      0     0       0     725.9    1061.4
            10290801-all-nlbi        ok      37      13     2     11     0       24       0      0     0       0     726.0    1062.4
UFDTNIA     10290801-allq-master     ok       1       1     0      1     1        0       0      0     0       0       0.4      13.2
            10290801-allq-nlbi       ok       1       1     0      1     0        0       0      0     0       0       0.4      13.0
UFDTNIRA    10290801-allq-master     ok    1839    1633     0   1633  1543       53       0      0     0       0    1985.8   21430.0
            10290801-allq-nlbi       ok    1839    1633     0   1633    90       54       0      0     0       0    1997.6   21414.0
UFFPDTNIRA  10290801-allq-master     ee      21       0     0      0     0        0       0     21     0       0       0.0      33.8
            10290801-allq-nlbi       ee      21       0     0      0     0        0       0     21     0       0       0.0      34.6
UFNIA       10290801-allq-master     ok   13509    5266  1229   4037  4715     3085       0      0    10       0  102597.1  969674.4
            10290801-allq-nlbi       ok   13509    5270  1232   4038   565     3078       0      0    14       0  102564.5  967828.1
```
The numbers changed slightly (note that the timeout was only 30s, thus introducing quite a bit of noise), but the story remains the same.

**Pull request merged**